File: app/src/main/java/com/tj/newsreader/ui/main/MainViewModel.java
package com.tj.newsreader.ui.main;

import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.LiveDataReactiveStreams;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModel;

import com.tj.newsreader.models.Article;
import com.tj.newsreader.network.main.MainApi;

import javax.inject.Inject;

import io.reactivex.Scheduler;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;



import static io.reactivex.internal.operators.single.SingleInternalHelper.toObservable;

public class MainViewModel extends ViewModel {

    private static final String TAG = "AuthViewModel";

    private final MainApi mainApi;

    //
    private MediatorLiveData<Article> mainArticle = new MediatorLiveData<>();

    @Inject
    public MainViewModel(MainApi mainApi) {
        this.mainApi = mainApi;
        Log.d(TAG, "MainViewModel:viewmodel is working...");

    }

    //Init call of the API

    public void authenticateWithtitle(String sources, String apiKey){

        final LiveData<Article> source = LiveDataReactiveStreams.fromPublisher(
                mainApi.getArticle(sources,apiKey)
                .subscribeOn(Schedulers.io())


        );

        mainArticle.addSource(source, new Observer<Article>() {
            @Override
            public void onChanged(Article article) {
                mainArticle.setValue(article);
                mainArticle.removeSource(source);
            }
        });





    }


    public LiveData<Article> observeArticle(){
        return mainArticle;


    }
}

File: app/src/main/java/com/tj/newsreader/ui/main/MainViewModel.kt
package com.tj.newsreader.ui.main

import android.util.Log

import androidx.lifecycle.LiveData
import androidx.lifecycle.LiveDataReactiveStreams
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModel

import com.tj.newsreader.models.Article
import com.tj.newsreader.network.main.MainApi

import javax.inject.Inject

import io.reactivex.Scheduler
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers


import io.reactivex.internal.operators.single.SingleInternalHelper.toObservable

class MainViewModel @Inject
constructor(private val mainApi: MainApi) : ViewModel() {

    //
    private val mainArticle = MediatorLiveData<Article>()

    init {
        Log.d(TAG, "MainViewModel:viewmodel is working...")

    }

    //Init call of the API

    fun authenticateWithtitle(sources: String, apiKey: String) {

        val source = LiveDataReactiveStreams.fromPublisher(
            mainApi.getArticle(sources, apiKey)
                .subscribeOn(Schedulers.io())


        )

        mainArticle.addSource(source) { article ->
            mainArticle.value = article
            mainArticle.removeSource(source)
        }


    }


    fun observeArticle(): LiveData<Article> {
        return mainArticle


    }

    companion object {

        private val TAG = "AuthViewModel"
    }
}
